File: MasteringSwift/MasteringSwift/chapters/chapter07/advanced/AdvancedCollections.swift
//
//  AdvancedCollections.swift
//  MasteringSwift
//
//  Capítulo 07 — Colecciones.
//  Nivel AVANZADO: Procesamiento de datos (calificaciones, productos).
//

import Foundation

/// Ejecuta ejemplos avanzados de colecciones. Llamar desde un test o playground.
func runAdvancedCollectionExamples() {
    // --- Sistema de calificaciones: diccionario de nombre a puntuación ---
    let grades: [String: Int] = [
        "Alice": 85,
        "Bob": 92,
        "Charlie": 78,
        "Diana": 95,
        "Eve": 88
    ]
    print("========================================")
    print("       STUDENT GRADES & PRODUCTS")
    print("========================================")
    print("\n--- Grades ---")
    print("Grades: \(grades)")

    let scores = Array(grades.values)
    let averageScore = Double(scores.reduce(0, +)) / Double(scores.count)
    print("Average score: \(averageScore)")

    let highScoreThreshold = 85
    let highScores = grades.filter { $0.value >= highScoreThreshold }
    print("Students with score >= \(highScoreThreshold): \(highScores)")

    // --- Lista de productos con precios: array de tuplas (nombre, precio) ---
    let products: [(String, Double)] = [
        ("Apple", 1.50),
        ("Banana", 0.80),
        ("Orange", 2.00),
        ("Milk", 3.20)
    ]
    print("\n--- Products ---")
    for (name, price) in products {
        print("  \(name): $\(price)")
    }
    let totalPrice = products.map { $0.1 }.reduce(0, +)
    let avgPrice = totalPrice / Double(products.count)
    print("Total price: $\(totalPrice)")
    print("Average price: $\(avgPrice)")

    // --- Filtrar productos caros (precio >= 2.0) ---
    let expensive = products.filter { $0.1 >= 2.0 }
    print("Products with price >= 2.0: \(expensive)")
    print("========================================")
}

File: MasteringDart/lib/chapters/chapter07/advanced/advanced_collections.dart
// ignore_for_file: avoid_print
//
//  advanced_collections.dart
//  MasteringDart
//
//  Capítulo 07 — Colecciones.
//  Nivel AVANZADO: Procesamiento de datos (calificaciones, productos).
//
//  Ejecutar con: dart run lib/chapters/chapter07/advanced/advanced_collections.dart

/// Ejecuta ejemplos avanzados de colecciones.
void runAdvancedCollectionExamples() {
  // --- Sistema de calificaciones: map de nombre a puntuación ---
  final grades = <String, int>{
    'Alice': 85,
    'Bob': 92,
    'Charlie': 78,
    'Diana': 95,
    'Eve': 88,
  };
  print('========================================');
  print('       STUDENT GRADES & PRODUCTS');
  print('========================================');
  print('\n--- Grades ---');
  print('Grades: $grades');

  final scores = grades.values;
  final averageScore =
      scores.reduce((a, b) => a + b) / scores.length;
  print('Average score: $averageScore');

  const highScoreThreshold = 85;
  final highScores =
      Map.fromEntries(grades.entries.where((e) => e.value >= highScoreThreshold));
  print('Students with score >= $highScoreThreshold: $highScores');

  // --- Lista de productos con precios (lista de registros o pares) ---
  final products = [
    ('Apple', 1.50),
    ('Banana', 0.80),
    ('Orange', 2.00),
    ('Milk', 3.20),
  ];
  print('\n--- Products ---');
  for (final item in products) {
    print('  ${item.$1}: \$${item.$2}');
  }
  final totalPrice = products.fold<double>(0, (sum, p) => sum + p.$2);
  final avgPrice = totalPrice / products.length;
  print('Total price: \$$totalPrice');
  print('Average price: \$$avgPrice');

  // --- Filtrar productos caros (precio >= 2.0) ---
  final expensive = products.where((p) => p.$2 >= 2.0).toList();
  print('Products with price >= 2.0: $expensive');
  print('========================================');
}

void main() {
  runAdvancedCollectionExamples();
}
